src/nnet3/nnet-simple-component.cc

```diff
@@ -295,6 +295,110 @@ void NoOpComponent::Backprop(const std::string &debug_info,
+  std::stringstream stream;
+  stream << Type() << ", dim=" << dim_
+         << ", norm-based-clipping=" << (norm_based_clipping_ ? "true" : "false")
+         << ", clipping-threshold=" << clipping_threshold_;
+  return stream.str();
+}
+
+void GradientClipperComponent::Init(int32 dim, 
+                                    BaseFloat clipping_threshold,
+                                    bool norm_based_clipping)  {
+  KALDI_ASSERT(clipping_threshold >= 0 && dim > 0);
+  dim_ = dim;
+  norm_based_clipping_ = norm_based_clipping;
+  clipping_threshold_ = clipping_threshold;
+}
+
+void GradientClipperComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 dim = 0;
+  bool ok = cfl->GetValue("dim", &dim);
+  bool norm_based_clipping = false;
+  BaseFloat clipping_threshold = 15.0;
+  cfl->GetValue("clipping-threshold", &clipping_threshold);
+  cfl->GetValue("norm-based-clipping", &norm_based_clipping);
+  if (!ok || cfl->HasUnusedValues() ||
+      clipping_threshold < 0 || dim <= 0)
+    KALDI_ERR << "Invalid initializer for layer of type "
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  Init(dim, clipping_threshold, norm_based_clipping);
+}
+
+void GradientClipperComponent::Propagate(
+                                 const ComponentPrecomputedIndexes *indexes,
+                                 const CuMatrixBase<BaseFloat> &in,
+                                 CuMatrixBase<BaseFloat> *out) const {
+  out->CopyFromMat(in);
+}
+
+void GradientClipperComponent::Backprop(const std::string &debug_info,
+                             const ComponentPrecomputedIndexes *indexes,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &out_deriv,
+                             Component *to_update, // may be NULL; may be identical
+                             // to "this" or different.
+                             CuMatrixBase<BaseFloat> *in_deriv) const {
+  // the following statement will do nothing if in_deriv and out_deriv have same
+  // memory.
+  in_deriv->CopyFromMat(out_deriv);
+  BaseFloat norm_before_clipping = 0,
+            norm_after_clipping = 0;
+
+  if (GetVerboseLevel() >= 2) {
+    norm_before_clipping = out_deriv.FrobeniusNorm();
+  }
+  if (clipping_threshold_ > 0) {
+    if (norm_based_clipping_) {
+      // each row in the derivative matrix, which corresponds to one sample in
+      // the mini-batch, is scaled to have a max-norm of clipping_threshold_
+      BaseFloat clipping_threshold_2 = pow(clipping_threshold_, 2);
+      CuVector<BaseFloat> deriv_norm(in_deriv->NumRows());
+      deriv_norm.AddDiagMat2(1.0 / in_deriv->NumCols(),
+                             *in_deriv, kNoTrans, 0.0);
+      deriv_norm.ApplyFloor(clipping_threshold_2);
+      deriv_norm.Scale(1.0 / clipping_threshold_2);
+      deriv_norm.ApplyPow(-0.5);
+      in_deriv->MulRowsVec(deriv_norm);
```

Comment:
On Mon, Aug 31, 2015 at 5:20 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/103#discussion_r38361083>:
>
> > +
> > +  if (GetVerboseLevel() >= 2) {
> > +    norm_before_clipping = out_deriv.FrobeniusNorm();
> > +  }
> > +  if (clipping_threshold_ > 0) {
> > +    if (norm_based_clipping_) {
> > +      // each row in the derivative matrix, which corresponds to one sample in
> > +      // the mini-batch, is scaled to have a max-norm of clipping_threshold_
> > +      BaseFloat clipping_threshold_2 = pow(clipping_threshold_, 2);
> > +      CuVector<BaseFloat> deriv_norm(in_deriv->NumRows());
> > +      deriv_norm.AddDiagMat2(1.0 / in_deriv->NumCols(),
> > +                             *in_deriv, kNoTrans, 0.0);
> > +      deriv_norm.ApplyFloor(clipping_threshold_2);
> > +      deriv_norm.Scale(1.0 / clipping_threshold_2);
> > +      deriv_norm.ApplyPow(-0.5);
> > +      in_deriv->MulRowsVec(deriv_norm);
>
> I recommend that you put something in the Info() call regarding printing
> the stats, so you can get it by running nnet3-info. This will get printed
> in the "progress" logs.
>
OK.


> You could also modify the nnet-training code so that instead of
> if (config.store_component_stats && config.zero_component_stats)
> ZeroComponentStats(nnet);
>
> it says
> if (config.zero_component_stats)
> ZeroComponentStats(nnet);
> so the stats will get zeroed before each iteration of training.
>
OK.


> It would also make sense to change the way the Scale() and Add() functions
> are declared, by moving them to the Component base-class and giving them,
> there, a default implementation that does nothing. This will allow you to
> simplify the calling code for Scale() and Add(). You can modify your
> GradientClipperComponent so that it overrides these functions, and uses
> them to scale/add the stats.
>
OK.


> Also, please rename from GradientClipperComponent to
> ClipGradientComponent: the other components have "imperative" names like
> SumGroupComponent or NormalizeComponent, I'd rather be consistent with them.
>
OK.



> Sorry for the inconvenience.
>
No problem, this helps me understand the interface better.


> Dan
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/103/files#r38361083>.
>